staging targets v3: Update RSA key log ID, reorder certificates

Hayden-IO opened the pull request:
The log ID should be computed over the PKIX encoded key, not PKCS1 encoded key. Also the certificate order should be reversed, intermediate to root.

Also included a staging targets directory and script to generate log IDs.

Steps to generate this:
* `go run cmd/generate_staging/key_to_log_id.go --key "<key value>"` to get log ID
* Add staging/targets folder, copy in previous targets, update trusted_root.json
* Manually update cmd/tuf/app/init.go to update expirations to 5+ years for all metadata
* `export TEST_KEY=./tests/test_data/cosign.key`
* `export TIMESTAMP_KEY=$TEST_KEY`
* `export SNAPSHOT_KEY=$TEST_KEY`
* `export TARGET_DIR=staging/targets`
* `export TARGET_META=staging-targets-metadata.yml`
* `export REPO=staging`
* `export GITHUB_USER=haydentherapper`
* `LOCAL=1 ./scripts/step-1.5.sh`
* Copy over the previously signed root and delegation file
* ./tuf sign -roles targets -key staging/keys/76651934/76651934_privkey.pem -repository staging
* ./tuf snapshot -repository staging
* ./tuf sign -roles snapshot -key tests/test_data/cosign.key -repository staging
* ./tuf timestamp -repository staging
* ./tuf sign -roles timestamp -key tests/test_data/cosign.key -repository staging
* ./tuf publish -repository staging

Fixes https://github.com/sigstore/root-signing/issues/768

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

Hayden-IO commented:
I'm not sure if I was using the scripts wrong, but is there a way to init the staged folder without initializing a new root?

Other thing was I had to manually delete the previous trusted_root files, not sure if this is a bug or if I misused something.

Hayden-IO commented:
Oh one other thing, how do we avoid re-signing new delegation metadata every time? I had to copy it over, otherwise the add delegations command would have me re-signing the metadata

Hayden-IO commented:
Going to update to fix https://github.com/sigstore/root-signing/issues/768 too

Hayden-IO commented:
```
$ cosign initialize --mirror=http://localhost:8001 --root=root-signing/staging/repository/1.root.json
Root status:
 {
	"local": "/usr/local/google/home/hblauzvern/.sigstore/root",
	"remote": "http://localhost:8001",
	"metadata": {
		"root.json": {
			"version": 2,
			"len": 1873,
			"expiration": "29 Sep 28 21:10 UTC",
			"error": ""
		},
		"snapshot.json": {
			"version": 3,
			"len": 1039,
			"expiration": "04 May 28 22:49 UTC",
			"error": ""
		},
		"targets.json": {
			"version": 3,
			"len": 4122,
			"expiration": "13 Oct 28 22:47 UTC",
			"error": ""
		},
		"timestamp.json": {
			"version": 3,
			"len": 719,
			"expiration": "27 Apr 28 22:49 UTC",
			"error": ""
		}
	},
	"targets": [
		"ctfe.pub",
		"ctfe_2022.pub",
		"ctfe_2022_2.pub",
		"fulcio.crt.pem",
		"fulcio_intermediate.crt.pem",
		"rekor.pub",
		"trusted_root.json"
	]
}
```

Root not updated, target/timestamp/snapshot are v3, as expected

kommendorkapten commented on staging/targets/trusted_root.json:
```diff
@@ -0,0 +1,86 @@
+{
```
I may be missing some context here, but this is not the latest trusted root (the TSA cert chain is missing). This might be the purpose though :)

Hayden-IO replied:
@kommendorkapten Is there an instance of the TSA in a GitHub staging environment? There's the values in prod which I could reuse, but I'd assume we would prefer a different instance.

kommendorkapten replied:
Yes it is. Let me find the certificate chain for that.

Hayden-IO replied:
@kommendorkapten Did you get a chance to find it? Otherwise I'll leave it blank for now.

kommendorkapten replied:
@haydentherapper sorry, was busy last Friday and was off until today (Public holiday here). Give me until the end of this day (Tuesday) and I should have it for you.

kommendorkapten replied:
@haydentherapper we are about to generate new staging keys this week, you want to wait for that or go on with an empty set for now?

Hayden-IO replied:
If I get a chance to this week I might continue with the generation just to make sure I have the steps, but in all likelihood I'll probably just wait til next week. Once you have the chain, can you either post it here or in an issue?

kommendorkapten replied:
I'll create an issue once we have the updated chain 👍

kommendorkapten commented:
```shell
./trtool verify -f ../root-signing/staging/
targets/trusted_root.json
Certificate's 'not before' must be before the CA's start time
error: verification failed
```

Looking deeper: 
Intermediate:
```
        Serial Number:
            26:d6:bf:a2:47:e0:1e:f8:da:6c:69:b5:04:ec:ee:86:bc:00:d5
    Signature Algorithm: ecdsa-with-SHA384
        Issuer: O=sigstore.dev, CN=sigstore
        Validity
            Not Before: Apr 14 21:38:40 2022 GMT
            Not After : Mar 22 16:50:45 2032 GMT
        Subject: O=sigstore.dev, CN=sigstore-intermediate
```
Root:
```
        Serial Number:
            0d:d1:09:be:5b:a5:88:4d:00:cd:86:88:13:88:d3:31:d9:c5:4f
    Signature Algorithm: ecdsa-with-SHA384
        Issuer: O=sigstore.dev, CN=sigstore
        Validity
            Not Before: Mar 25 16:50:46 2022 GMT
            Not After : Mar 22 16:50:45 2032 GMT
        Subject: O=sigstore.dev, CN=sigstore
```

And the valid for in the `trusted_root.json`:
```
      "validFor": {
        "start": "2022-03-25T16:50:46.000Z"
      }
```

The valid for should point out when the entire chain was valid, and so be the smallest interval of when all certificates are valid.

kommendorkapten commented on staging/targets/trusted_root.json:
```diff
@@ -0,0 +1,86 @@
+        ]
+      },
+      "validFor": {
+        "start": "2022-03-25T16:50:46.000Z"
```
```suggestion
        "start": "2022-04-14T21:38:40.000Z"
```

Hayden-IO replied:
Will update, thanks! Is trtool open sourced?

kommendorkapten commented:
> I'm not sure if I was using the scripts wrong, but is there a way to init the staged folder without initializing a new root?

~Not sure I follow your intention here?~

You just want to update `targets.json`, not a new root version?

> Oh one other thing, how do we avoid re-signing new delegation metadata every time? I had to copy it over, otherwise the add delegations command would have me re-signing the metadata

This is a good question. Not sure tbh, and I'm thinking if this is a scenario we would end up in in the future? I would expect in the future when a new root is staged, the existing (signed) delegation would be kept?

Not strictly related though but we set the expiration of the delegation to the same as the root/target, so for next root signing we would need to sign the delegation again.

asraa commented:
> I'm not sure if I was using the scripts wrong, but is there a way to init the staged folder without initializing a new root?

There WAS a change in the root (expiration) - each time we do a root signing though we expect to refresh root expiration.

If you make a change to a file that's not the root, the staged folder will initialize. For e.g. when you run `./tuf snapshot` it initializes. You just want to refresh the targets in this case - so technically you need to separate the commands. They're bundled in our repo because they're both HSM signed so we want to reduce the # of times we ask HSM signers to sign.

asraa commented:
> Oh one other thing, how do we avoid re-signing new delegation metadata every time? I had to copy it over, otherwise the add delegations command would have me re-signing the metadata

If you want to keep the delegation the same, don't run add-delegation again. But I think in this case, you had modified the expiration? I'm not sure what you did here, I'd probably have to watch. You don't have to run add-delegation again once it's added and you're not modifying it.

asraa commented:
You shouldn't have had to do this step:

> Add staging/targets folder, copy in previous targets, update trusted_root.json

That may have caused some issues - ideally you just modify the targets configuration file and the targets in the targets/ subfolder. And then run the ceremony steps.

It would have incremented the root, but is that an issue? (Reason is the above - we bundle those changes so we can bundle HSM sigs in prod)

Hayden-IO commented:
> If you want to keep the delegation the same, don't run add-delegation again. But I think in this case, you had modified the expiration? I'm not sure what you did here, I'd probably have to watch. You don't have to run add-delegation again once it's added and you're not modifying it.

Hm, I'll try again, but when I ran init, it didn't copy the delegation in. Though I realize that might be because I had two separate metadata yamls, one for delegates and one for the primary targets. If I merge these together, should init work properly?

> It would have incremented the root, but is that an issue? (Reason is the above - we bundle those changes so we can bundle HSM sigs in prod)

No particular reason, I can regenerate this with a root update, it definitely simplifies things.

kommendorkapten commented:
Yes!
https://github.com/kommendorkapten/trtool

It’s still very early with limited functionality, but have some ideas for
future improvements.

On Fri, 14 Apr 2023 at 18:02, Hayden B ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In staging/targets/trusted_root.json
> <https://github.com/sigstore/root-signing/pull/799#discussion_r1167020259>
> :
>
> > +        "organization": "sigstore.dev",
> +        "commonName": "sigstore"
> +      },
> +      "uri": "https://fulcio.sigstage.dev",
> +      "certChain": {
> +        "certificates": [
> +          {
> +            "rawBytes": "MIICGTCCAaCgAwIBAgITJta/okfgHvjabGm1BOzuhrwA1TAKBggqhkjOPQQDAzAqMRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIyMDQxNDIxMzg0MFoXDTMyMDMyMjE2NTA0NVowNzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRlcm1lZGlhdGUwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAASosAySWJQ/tK5r8T5aHqavk0oI+BKQbnLLdmOMRXHQF/4Hx9KtNfpcdjH9hNKQSBxSlLFFN3tvFCco0qFBzWYwZtsYsBe1l91qYn/9VHFTaEVwYQWIJEEvrs0fvPuAqjajezB5MA4GA1UdDwEB/wQEAwIBBjATBgNVHSUEDDAKBggrBgEFBQcDAzASBgNVHRMBAf8ECDAGAQH/AgEAMB0GA1UdDgQWBBRxhjCmFHxib/n31vQFGn9f/+tvrDAfBgNVHSMEGDAWgBT/QjK6aH2rOnCv3AzUGuI+h49mZTAKBggqhkjOPQQDAwNnADBkAjAM1lbKkcqQlE/UspMTbWNo1y2TaJ44tx3l/FJFceTSdDZ+0W1OHHeU4twie/lq8XgCMHQxgEv26xNNiAGyPXbkYgrDPvbOqp0UeWX4mJnLSrBr3aN/KX1SBrKQu220FmVL0Q=="
> +          },
> +          {
> +            "rawBytes": "MIIB9jCCAXugAwIBAgITDdEJvluliE0AzYaIE4jTMdnFTzAKBggqhkjOPQQDAzAqMRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIyMDMyNTE2NTA0NloXDTMyMDMyMjE2NTA0NVowKjEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MREwDwYDVQQDEwhzaWdzdG9yZTB2MBAGByqGSM49AgEGBSuBBAAiA2IABMo9BUNk9QIYisYysC24+2OytoV72YiLonYcqR3yeVnYziPt7Xv++CYE8yoCTiwedUECCWKOcvQKRCJZb9ht4Hzy+VvBx36hK+C6sECCSR0x6pPSiz+cTk1f788ZjBlUZaNjMGEwDgYDVR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFP9CMrpofas6cK/cDNQa4j6Hj2ZlMB8GA1UdIwQYMBaAFP9CMrpofas6cK/cDNQa4j6Hj2ZlMAoGCCqGSM49BAMDA2kAMGYCMQD+kojuzMwztNay9Ibzjuk//ZL5m6T2OCsm45l1lY004pcb984L926BowodoirFMcMCMQDIJtFHhP/1D3a+M3dAGomOb6O4CmTry3TTPbPsAFnv22YA0Y+P21NVoxKDjdu0tkw="
> +          }
> +        ]
> +      },
> +      "validFor": {
> +        "start": "2022-03-25T16:50:46.000Z"
>
> Will update, thanks! Is trtool open sourced?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sigstore/root-signing/pull/799#discussion_r1167020259>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACSDVRAN2EZJISNG7C7AXY3XBFYKXANCNFSM6AAAAAAW5UYEFQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

asraa commented:
> Hm, I'll try again, but when I ran init, it didn't copy the delegation in.

Are you sure? 

If there are no changes to a delegation, it won't end up in the `staging` folder.

Hayden-IO commented:
I guess the root of this is that I have metadata that is long-lived, unlike the production metadata where the delegation metadata will expire at the same time as the root/targets, so everything needs to be re-signed. So maybe I'm overthinking this, and once we have some automation around staging, this'll be resolved.

asraa commented:
I think it's resolved in both cases! I don't think there's an issue - it's just that if the delegaiton metadata was unchanged, it doesn't appear in the staging subfolder when you run init. Snapshotting will still snapshot it.

Hayden-IO commented:
Tested the GCS bucket sync, finally now working with workload identity. Will clean up this PR

Hayden-IO commented:
Finally getting around to fixing this

Hayden-IO commented:
Staging TSA:

```
"timestampAuthorities": [
    {
      "subject": {
        "organization": "GitHub, Inc.",
        "commonName": "Internal Services Root - staging"
      },
      "uri": "tsa.github.internal",
      "certChain": {
        "certificates": [
          {
            "rawBytes": "MIICEzCCAZigAwIBAgIUMpRykCcZaSOBipYce6r2DlnM7vUwCgYIKoZIzj0EAwMwPDEVMBMGA1UEChMMR2l0SHViLCBJbmMuMSMwIQYDVQQDExpUU0EgaW50ZXJtZWRpYXRlIC0gc3RhZ2luZzAeFw0yMzA2MTQxMjAwMDBaFw0yNDA2MTMxMjAwMDBaMDwxFTATBgNVBAoTDEdpdEh1YiwgSW5jLjEjMCEGA1UEAxMaVFNBIFRpbWVzdGFtcGluZyAtIHN0YWdpbmcwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATWAMg1BEHAzb03PHUKJiRJZdXcKIL0K/ks3Ylq5F5YDRIxUN4o8yeIaCWXa6i16zi8nXMFsa+3XrYM8mUyi9F6o3gwdjAOBgNVHQ8BAf8EBAMCB4AwDAYDVR0TAQH/BAIwADAdBgNVHQ4EFgQUr+RZdcTNo31p3FuR0pajelcnr40wHwYDVR0jBBgwFoAUCmpQzrB6hAnlizGx37LrhKEzsT4wFgYDVR0lAQH/BAwwCgYIKwYBBQUHAwgwCgYIKoZIzj0EAwMDaQAwZgIxAIPUlZB2/p5rpCM3HCn1R8G5TIIW6aZPKtfPWDkNQY0bpFu6e8Dkm2TV3jfgYExuBgIxAOA+vTlDDJoz/qTMMs8VSpw3AMgktlMEhd8V0E+aLW5OizfphuiidkqkqkbCwRSW1w=="
          },
          {
            "rawBytes": "MIICNzCCAb6gAwIBAgIUUXRsBKgGXjrJbdJCQPJMsjfyJcYwCgYIKoZIzj0EAwMwQjEVMBMGA1UEChMMR2l0SHViLCBJbmMuMSkwJwYDVQQDEyBJbnRlcm5hbCBTZXJ2aWNlcyBSb290IC0gc3RhZ2luZzAeFw0yMzA2MTQwMDAwMDBaFw0yODA2MTIwMDAwMDBaMDwxFTATBgNVBAoTDEdpdEh1YiwgSW5jLjEjMCEGA1UEAxMaVFNBIGludGVybWVkaWF0ZSAtIHN0YWdpbmcwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAAS/upihrvu/9+w1Ybfog3B1a8KfQa1bn7DLcJz2iumo+oCfg2bcbyRWygu8zRmrzJKp4HgQHC4LZJEEWm/MNIN1o6wVVmiDTZw01tk4aInmRVF13VKMscdzW5Ho4sYaeOejezB5MA4GA1UdDwEB/wQEAwIBBjATBgNVHSUEDDAKBggrBgEFBQcDCDASBgNVHRMBAf8ECDAGAQH/AgEAMB0GA1UdDgQWBBQKalDOsHqECeWLMbHfsuuEoTOxPjAfBgNVHSMEGDAWgBR04GYtT79vaQmAEPPEte8q+W1KRTAKBggqhkjOPQQDAwNnADBkAjBOTWZP1QYnmHpFqL73eSzhmSLiHs9pXsQghK0p8pvlAg0R9bxAyXGIZ8qx+k2iIGcCMDRQIScz0gu2Xef3++p2vFYouBsIKbqxv0raJuIlmGiYEvb22MDpAitevKAgqNVMEg=="
          },
          {
            "rawBytes": "MIICCDCCAY6gAwIBAgIULHHP/UhbXJbdyZfT6gHgTzIYF4EwCgYIKoZIzj0EAwMwQjEVMBMGA1UEChMMR2l0SHViLCBJbmMuMSkwJwYDVQQDEyBJbnRlcm5hbCBTZXJ2aWNlcyBSb290IC0gc3RhZ2luZzAeFw0yMzA2MTQwMDAwMDBaFw0zMzA2MTEwMDAwMDBaMEIxFTATBgNVBAoTDEdpdEh1YiwgSW5jLjEpMCcGA1UEAxMgSW50ZXJuYWwgU2VydmljZXMgUm9vdCAtIHN0YWdpbmcwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAASocByKBUdzgtqRXcpe/AE5oPoDMWTQqz1/jUQOA8qoEjYBXg9gfGU5KHK/UdwQc4lxbZEA9nJS9vUQAMVV5Es9B4thNHThKR4hFmCL8kKIEoMzXx282Qr6x4ZHYk4tQsCjRTBDMA4GA1UdDwEB/wQEAwIBBjASBgNVHRMBAf8ECDAGAQH/AgECMB0GA1UdDgQWBBR04GYtT79vaQmAEPPEte8q+W1KRTAKBggqhkjOPQQDAwNoADBlAjBBOu3RtlH1FLvfHPhyoWJHgm+PSNrsWQLRkmWQgAfNPYsfO5fWyhAebMV3FpKVPBICMQCVaie4NAsGi+AHLDhGnPn4Qptz0LBH2So6AVJS24ICeDUDQxKeUTNkUsy6Qgg97/4="
          }
        ]
      },
      "validFor": {
        "start": "2023-06-15T00:00:00Z"
      }
    }
  ]





```

Hayden-IO committed:
Staging: Update RSA key ID and certificate order

The log ID should be computed over the PKIX encoded key, not PKCS1
encoded key. The certificates should also be ordered in reverse.

Also included a staging targets directory and script to generate log
IDs.

Signed-off-by: Hayden Blauzvern <hblauzvern@google.com>

Hayden-IO commented:
@kommendorkapten @asraa This is _finally_ ready for review, sorry for the very long delay!

Hayden-IO commented:
This adds the github staging TSA certs and fixes startTime.

Hayden-IO commented:
fyi @jku, if you're working with staging metadata, this is an update. this should be the last update

kommendorkapten reviewed (approved):
Looks good. 

Also verified `trusted_root.json`:
```shell
$ ./trtool verify -v -f hayden.tr.json
Verifying OU='sigstore.dev' CN='sigstore' of length 2
  Loaded OU='sigstore.dev' CN='sigstore-intermediate' CA:true MaxPathLen 0 at pos 0
    issuer OU='sigstore.dev' CN='sigstore'
  Loaded OU='sigstore.dev' CN='sigstore' CA:true MaxPathLen -1 at pos 1
    issuer OU='sigstore.dev' CN='sigstore'
------------------------------------------------------------------------
Verifying OU='GitHub, Inc.' CN='Internal Services Root - staging' of length 3
  Loaded OU='GitHub, Inc.' CN='TSA Timestamping - staging' CA:false MaxPathLen -1 at pos 0
    issuer OU='GitHub, Inc.' CN='TSA intermediate - staging'
  Loaded OU='GitHub, Inc.' CN='TSA intermediate - staging' CA:true MaxPathLen 0 at pos 1
    issuer OU='GitHub, Inc.' CN='Internal Services Root - staging'
  Loaded OU='GitHub, Inc.' CN='Internal Services Root - staging' CA:true MaxPathLen 2 at pos 2
    issuer OU='GitHub, Inc.' CN='Internal Services Root - staging'
------------------------------------------------------------------------
Trusted root is valid
```
